ocaml/xapi/test_network_event_loop.ml

```diff
@@ -0,0 +1,165 @@
+let test_network_event_loop ~no_nbd_networks_at_start () =
+  let __context, _ = Test_event_common.event_setup_common () in
+  (* We need to set test_mode to true to ensure that the event loop will use
+     the local Xapi_events.from function instead of the Client module, which
+     would cause the test to fail. *)
+  Helpers.test_mode := true;
+
+  let localhost = Helpers.get_localhost ~__context in
+  let other_host = Test_common.make_host ~__context () in
+
+  (* We have to wait for a bit for the event loop to notice the changes, without a delay the test will fail. *)
+  let delay = 0.2 in
+  let network_event_loop_wait_after_failure_seconds = 0.4 in
+  let received_params = ref None in
+
+  (* We simulate failure of the firewall update script this way *)
+  let fail_firewall_update = ref false in
+
+  let start_event_loop = Thread.create
+      (fun () ->
+         Network_event_loop._watch_networks
+           __context
+           ~update_firewall:(fun pifs ->
+               if !fail_firewall_update then failwith "Failed to update firewall";
+               received_params := Some pifs
+             )
+           ~wait_after_failure_seconds:network_event_loop_wait_after_failure_seconds
+      )
+  in
+
+  let assert_received_params msg expected =
+    Thread.delay delay;
+    match !received_params with
+    | None -> OUnit.assert_failure ("The update_firewall function was not called: " ^ msg)
+    | Some p ->
+      Ounit_comparators.StringSet.(assert_equal (of_list expected) (of_list p))
```

Comment:
The first item has to be the same string ("set") in each list: just checking the lists contain the same values is not enough.
So maybe add a line or two to check that if the expected list has at least one value then the first values match.